### PR TITLE
Ignore coroutines state machines

### PR DESCRIPTION
**Pull request by zuevmaxim**

### CODE REVIEW

File: instrumentation/src/com/intellij/rt/coverage/instrumentation/LineEnumerator.java [hunk 1]
```diff
@@ -189,16 +188,17 @@ public void visitLookupSwitchInsn(Label dflt, int[] keys, Label[] labels) {
...
+        rememberSwitchLabels(dflt, labels);
+        SwitchData switchData = lineData.addSwitch(myCurrentSwitch++, min, max);
+        mySwitchLabels.put(dflt, switchData);
+      }
+      myState = SEEN_NOTHING;
+      myHasInstructions = true;
     }
-    myState = SEEN_NOTHING;
-    myHasInstructions = true;
+    super.visitTableSwitchInsn(min, max, dflt, labels);
```
Comment (gorrus) [inline instrumentation/src/com/intellij/rt/coverage/instrumentation/LineEnumerator.java:201]: it seems counter-intuitive that we call super at the end here, but not in other methods

File: src/com/intellij/rt/coverage/data/JumpsAndSwitches.java [hunk 2]
```diff
@@ -69,6 +69,12 @@ public SwitchData addSwitch(final int switchNumber, final int[] keys) {
     return mySwitches.get(switchNumber);
   }
 
+  public void removeSwitch(int switchNumber) {
+    if (mySwitches != null && 0 <= switchNumber && switchNumber < mySwitches.size()) {
```
Comment (gorrus) [inline src/com/intellij/rt/coverage/data/JumpsAndSwitches.java:73]: Is it really possible that mySwitches could be null here? Or switchNumber is out of bounds? Maybe it should not silently skip in such situations?
Same for removeJump

File: instrumentation/src/com/intellij/rt/coverage/instrumentation/filters/KotlinCoroutinesFilter.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  public void visitFieldInsn(int opcode, String owner, String name, String descriptor) {
+    super.visitFieldInsn(opcode, owner, name, descriptor);
+    // ignore return Unit line
+    myHasExecutableCode |= !(owner.equals("kotlin/Unit")
+        && name.equals("INSTANCE")
+        && descriptor.equals("Lkotlin/Unit;"));
+    myStateLabelVisited = opcode == Opcodes.GETFIELD
+        && name.equals("label")
+        && descriptor.equals("I")
+        && owner.startsWith(getClassName());
```
Comment (gorrus) [inline instrumentation/src/com/intellij/rt/coverage/instrumentation/filters/KotlinCoroutinesFilter.java:133]: Better do not do this yourself, use org.jetbrains.org.objectweb.asm.Type.getObjectType(owner).getClassName()

File: instrumentation/src/com/intellij/rt/coverage/instrumentation/LineEnumerator.java (hunk 1 above)
Comment (zuevmaxim) [reply instrumentation/src/com/intellij/rt/coverage/instrumentation/LineEnumerator.java]: I fixed filters order, so `super` call position does not depend on filters. Anyway `super` calls in `visitJump` are still unusual as we have to substitute labels. The same reason will lead to `super` call disorder in switch methods after #34 merge

**Review by gorrus — APPROVED** [review]

**Pull request merged**